[SPARK-39083][CORE] Fix race condition between update and clean app data

tanvn opened the pull request:
### What changes were proposed in this pull request?
make `cleanAppData` atomic to prevent race condition between update and clean app data. 
When the race condition happens, it could lead to a scenario when `cleanAppData` delete the entry of 
ApplicationInfoWrapper for an application right after it has been updated by `mergeApplicationListing`.
So there will be cases when the HS Web UI displays `Application not found` for applications whose logs does exist.

#### Error message
```
22/04/29 17:16:21 DEBUG FsHistoryProvider: New/updated attempts found: 1 ArrayBuffer(viewfs://iu/log/spark3/application_1651119726430_138107_1)
22/04/29 17:16:21 INFO FsHistoryProvider: Parsing viewfs://iu/log/spark3/application_1651119726430_138107_1 for listing data...
22/04/29 17:16:21 INFO FsHistoryProvider: Looking for end event; skipping 10805037 bytes from viewfs://iu/log/spark3/application_1651119726430_138107_1...
22/04/29 17:16:21 INFO FsHistoryProvider: Finished parsing viewfs://iu/log/spark3/application_1651119726430_138107_1
22/04/29 17:16:21 ERROR Utils: Uncaught exception in thread log-replay-executor-7
java.util.NoSuchElementException
        at org.apache.spark.util.kvstore.InMemoryStore.read(InMemoryStore.java:85)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$checkAndCleanLog$3(FsHistoryProvider.scala:927)
        at scala.Option.foreach(Option.scala:407)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$checkAndCleanLog$1(FsHistoryProvider.scala:926)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at org.apache.spark.util.Utils$.tryLog(Utils.scala:2032)
        at org.apache.spark.deploy.history.FsHistoryProvider.checkAndCleanLog(FsHistoryProvider.scala:916)
        at org.apache.spark.deploy.history.FsHistoryProvider.mergeApplicationListing(FsHistoryProvider.scala:712)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$checkForLogs$15(FsHistoryProvider.scala:576)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```
#### Background
Currently, the HS runs the `checkForLogs` to build the application list based on the current contents of the log directory for every 10 seconds by default.
- https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L296-L299
- https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L472

In each turn of execution, this method scans the specified logDir and parse the log files to update its KVStore:
- detect new updated/added files to process : https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L574-L578
- detect stale data to remove: https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L586-L600

These 2 operations are executed in different threads as `submitLogProcessTask` uses `replayExecutor` to submit tasks.
https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L1389-L1401

### When does the bug happen?
`Application not found` error happens in the following scenario:
In the first run of `checkForLogs`, it detected a newly-added log `viewfs://iu/log/spark3/AAA_1.inprogress` (log of an in-progress application named AAA). So it will add 2 entries to the KVStore
- one entry of key-value as the key is the logPath (`viewfs://iu/log/spark3/AAA_1.inprogress`)  and the value is an instance of LogInfo represented the log
  -  https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L495-L505
  - https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L545-L552
- one entry of key-value as the key is the applicationId (`AAA`) and  the value is an instance of ApplicationInfoWrapper holding the information of the application.
  - https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L825
  - https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L1172

In the next run of `checkForLogs`, now the AAA application has finished, the log `viewfs://iu/log/spark3/AAA_1.inprogress` has been deleted and a new log `viewfs://iu/log/spark3/AAA_1` is created. So  `checkForLogs` will do the following 2 things in 2 different threads:
- Thread 1: parsing the new log `viewfs://iu/log/spark3/AAA_1` and update data in its KVStore
  - add a new entry of key: `viewfs://iu/log/spark3/AAA_1` and value: an instance of LogInfo represented the log
  - updated the entry with key=applicationId (`AAA`)  with new value of  an instance of ApplicationInfoWrapper (for example: the isCompleted flag now change from false to true)
- Thread 2:  data related to `viewfs://iu/log/spark3/AAA_1.inprogress` is now considered as stale and it must be deleted.
  - clean App data for `viewfs://iu/log/spark3/AAA_1.inprogress` https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L586-L600
  - Inside `cleanAppData`, first it loads the latest information of `ApplicationInfoWrapper` from the KVStore: https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L632  For most of the time, when this line is executed, Thread 1 already finished `updating the entry with key=applicationId (AAA)  with new value of  an instance of ApplicationInfoWrapper` so this condition https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L637 will be evaluated as false, so `isStale` will be false. However, in some rare cases, when Thread1 does not finish the update yet, the old data of ApplicationInfoWrapper will be load, so `isStale` will be true and it leads to deleting the entry of ApplicationInfoWrapper in KVStore: https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L656-L662 and the worst thing is it delete the entry right after when Thread 1 has finished updating the entry with key=applicationId (`AAA`)  with new value of  an instance of ApplicationInfoWrapper. So the entry for the ApplicationInfoWrapper of applicationId= `AAA` is removed forever then when users access the Web UI for this application, and `Application not found` is shown up while the log for the app does exist.

So here we make the `cleanAppData` method atomic just like the `addListing` method https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L1172 so that
- If Thread 1 gets the lock on the `listing` before Thread 2, it will update the entry for the application, so in Thread2 `isStale` will be false, the entry for the application will not be removed from KVStore
- If Thread 2 gets the lock on the `listing` before Thread 1, then `isStale` will be true, the entry for the application will be removed from KVStore but after that it will be added again by Thread 1.
In both case, the entry for the application will not be deleted unexpectedly from KVStore.


### Why are the changes needed?
Fix the bug causing HS Web UI to display `Application not found` for applications whose logs does exist.


### Does this PR introduce _any_ user-facing change?
Yes, bug fix.


## How was this patch tested?
Manual test.
Deployed in our Spark HS and the `java.util.NoSuchElementException` exception does not happen anymore.
`Application not found` error does not happen anymore.

tanvn committed:
SPARK-39083 : fix race condition between update and clean app data

tanvn committed:
SPARK-39083: add comment (to trigger Github Action)

AmplabJenkins commented:
Can one of the admins verify this patch?

tanvn commented:
@gengliangwang @srowen @dongjoon-hyun @turboFei @vanzin @HeartSaVioR
Please take a look when you have time.

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -630,41 +630,44 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+            // in progress application. Just return that the app should be left alone.
+            false
+          } else {
+            val maybeUI = synchronized {
```
The nested synchronized makes me uneasy - risks a deadlock. Why not just `synchronized` on everything? safe, but is it too slow?

dongjoon-hyun replied:
+1 for @srowen 's advice.

tanvn committed:
SPARK-39083: remove nested synchronized

gengliangwang commented:
```
If Thread 1 gets the lock on listing before Thread 2, isStale will be false, the app will not be removed from KVStore
If Thread 2 gets the lock on listing before Thread 1, then isStale will be true, the app will be removed from KVStore but after that it will be added again by Thread 1.
In both case, the entry for the application will not be deleted unexpectedly from KVStore.
```
@tanvn It seems that the description doesn't match the current implementation. Can you update it?

tanvn commented:
> Can you update it?

@gengliangwang just updated the description. PTAL, thank you.

mridulm commented:
+CC @thejdeep, @shardulm94

tanvn commented:
@srowen @gengliangwang @mridulm 
JFYI, I have deployed this patch on my server 3 days ago and the `java.util.NoSuchElementException` Exception occurred 3 times. So I guess  just using `synchronized` on the code block of `cleanAppData` is not enough. And we should use `synchronized` on `listing` for a better understanding since it is being used in other places. WDYT?

gengliangwang commented:
@tanvn Thanks for trying the new patch. Yes the current approach doesn't fix all the race condition.
I take a look at the description again. The method `isProcessing` should only return false when all the updates are done. I wonder why the filtering doesn't work in https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L593

gengliangwang commented:
cc @jixiongxiao-db @rednaxelafx

tanvn commented:
@gengliangwang 
Thank you for taking your time.
Regarding `isProcessing ` method, in the scenario I wrote in the PR summary:
In the next run of checkForLogs, now the AAA application has finished, the log `viewfs://iu/log/spark3/AAA_1.inprogress` has been deleted and a new log `viewfs://iu/log/spark3/AAA_1` is created. 
In this case, `viewfs://iu/log/spark3/AAA_1` will be marked as `processing`.
https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L1391
but the `stale` is taking data from the `listing` map, 
https://github.com/apache/spark/blob/v3.2.1/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L586-L592
which means it will contain `viewfs://iu/log/spark3/AAA_1.inprogress` (as it is added to the `listing` in the previous execution of `checkForLogs`) and `isProcessing` for `viewfs://iu/log/spark3/AAA_1.inprogress` will return false, so `cleanAppData` will be executed for it (which I think there is no problem)

tanvn committed:
SPARK-39083: synchronized on listing when loading and removing app

tanvn commented:
@gengliangwang cc @srowen 
I just added some changes to the `cleanAppData` method, I think getting a lock on `listing` when loading and removing the stale app will fix the race condition and also I have moved out the code block for removing from `activeUIs` to avoid nested `synchronized`. PTAL when you have time.

mridulm commented:
Thanks for the test results @tanvn.
@thejdeep, can you check if this is an issue in our deployment ? And if yes, what we have done to mitigate it (if any) - else, any thoughts on how to address it ?

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -631,37 +631,38 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-          listing.write(newAppInfo)
-        } else {
-          listing.delete(classOf[ApplicationInfoWrapper], appId)
+      if(isStale) {
```
Nit: space after if
This looks OK - finer-grained but not nesting locks now.

tanvn replied:
Thank you. Just fixed it.

tanvn committed:
SPARK-39083: add space after if

tanvn commented:
@mridulm @gengliangwang @srowen @thejdeep 
JFYI, I have deployed the newest version of this patch on my server from yesterday and there is no exception so far. 
So I think the bug has been fixed and `synchronized` on the `listing` is the point (since the race condition is happening between the threads adding and removing data on `listing`).

srowen commented:
Merged to master/3.3/3.2
